ONSdigital/census-rh-service
# Cr129 support survey launched endpoint and send message to case service

# Motivation and Context
This change supports the receipt of an http request which states that a survey has been launched. This results in a SurveyLaunched message being written to a RabbitMQ exchange.

# What has changed
- This change has resulted in the creation of a RestController class (RespondentHomeEndpoint.java) to implement the "/SurveyLaunched" endpoint.
- DTO classes have also been created, to hold both the initial request and the content of the message which is written to the RabbitMQ exchange.
- Unit and integration tests have been added.

# How to test?
To test automatically run the unit and integration tests.

# How to manually test?
1 - In the RH project Edit src/main/resources/springintegration/broker.xml
Find the "\<!-- Start of Queues --\>"
Add the following to define the queues:
>>>
	<rabbit:queue name="Case.SurveyLaunched" durable="true">
		<rabbit:queue-arguments value-type="java.lang.String">
			<entry key="x-dead-letter-exchange" value="case-deadletter-exchange" />
			<entry key="x-dead-letter-routing-key" value="Case.Gateway.binding" />
		</rabbit:queue-arguments>
	</rabbit:queue>

 	<rabbit:queue name="Case.SurveyLaunchedDLQ" durable="true" />

2 - Then in the same file. Find the '\<!-- Start of Exchanges --\>' section.
Add to the bindings for the 'events' exchange.
`
	<rabbit:binding queue="Case.SurveyLaunched" key="event.response.authentication" />
`

3 - Create file with request body:
`
cat > /tmp/surveyLaunchedDTO.json << EOF
{
  "questionnaireId": "23434234234",
  "caseId": "3fa85f64-5717-4562-b3fc-2c963f66afa6"
}
EOF
`

4 - Start the RH server and submit a survey Launched post request:
``
$ curl -s -X POST -H "Content-Type: application/json" --data @/tmp/surveyLaunchedDTO.json --user admin:secret "http://localhost:8171/surveyLaunched"
``

5 - Check the content of the Case.surveyLaunched queue.

# Links
See CR: https://collaborate2.ons.gov.uk/jira/browse/CR-129
